File: 77.组合.py
#
# @lc app=leetcode.cn id=77 lang=python3
#
# [77] 组合
#
from typing import List

# @lc code=start
class Solution:
    def combine(self, n: int, k: int) -> List[List[int]]:
        assert n >= k
        if not n or not k:
            return []
        res = []
        self.backtrack(res, [], list(range(1, n+1)), k)
        return res

    def backtrack(self, res, path, remain, k):
        if not k:
            res.append(path[:])
            return
        for i, n in enumerate(remain):
            path.append(n)
            self.backtrack(res, path, remain[i+1:], k-1)
            path.pop()
            

# @lc code=end
# s = Solution()
# print(s.combine(3, 2))

File: 77.组合.java
/*
 * @lc app=leetcode.cn id=77 lang=java
 *
 * [77] 组合
 */
import java.util.*;

// @lc code=start
class Solution {
    public List<List<Integer>> combine(int n, int k) {
        List<List<Integer>> res = new ArrayList<>();
        if (k <= 0 || n < k) return res;
        Stack<Integer> track = new Stack<>();
        backtrack(n, k, 1, track, res);
        return res;
    }

    private void backtrack(int n, int k, int start, Stack<Integer> track, List<List<Integer>> res) {
        if (track.size() == k) {
            res.add(new ArrayList<>(track));     // make a copy of current path
            return;
        }
        for (int i = start; i <= n; ++i) {
            track.push(i);
            backtrack(n, k, i + 1, track, res);
            track.pop();
        }
    }
}
// @lc code=end

